sampling/include/var_opt_sketch_impl.hpp

```diff
@@ -0,0 +1,1591 @@
+
+  return os;
+}
+
+template<typename T, typename S, typename A>
+std::ostream& var_opt_sketch<T,S,A>::items_to_stream(std::ostream& os) const {
+  os << "### Sketch Items" << std::endl;
+
+  uint32_t print_length = (n_ < k_ ? n_ : k_ + 1);
+  for (int i = 0; i < print_length; ++i) {
+    if (i == h_) {
+      os << i << ": GAP" << std::endl;
+    } else {
+      os << i << ": " << data_[i] << "\twt = " << weights_[i] << std::endl;
+    }
+  }
+
+  return os;
+}
+
+template <typename T, typename S, typename A>
+std::string var_opt_sketch<T,S,A>::to_string() const {
+  std::ostringstream ss;
+  to_stream(ss);
+  return ss.str();
+}
+
+template <typename T, typename S, typename A>
+std::string var_opt_sketch<T,S,A>::items_to_string() const {
+  std::ostringstream ss;
+  items_to_stream(ss);
+  return ss.str();
+}
+
+template<typename T, typename S, typename A>
+void var_opt_sketch<T,S,A>::update(const T& item, double weight, bool mark) {
+  if (weight <= 0.0) { 
+    throw std::invalid_argument("Item weights must be strictly positive. Found: "
+                                + std::to_string(weight));
+  }
+  ++n_;
+
+  if (r_ == 0) {
+    // exact mode
+    update_warmup_phase(item, weight, mark);
+  } else {
+    // sketch is in estimation mode so we can make the following check
+    assert(h_ == 0 || (peek_min() >= get_tau()));
+
+    // what tau would be if deletion candidates turn out to be R plus the new item
+    // note: (r_ + 1) - 1 is intentional
+    double hypothetical_tau = (weight + total_wt_r_) / ((r_ + 1) - 1);
+
+    // is new item's turn to be considered for reservoir?
+    double condition1 = (h_ == 0) || (weight <= peek_min());
+
+    // is new item light enough for reservoir?
+    double condition2 = weight < hypothetical_tau;
+  
+    if (condition1 && condition2) {
+      update_light(item, weight, mark);
+    } else if (r_ == 1) {
+      update_heavy_r_eq1(item, weight, mark);
+    } else {
+      update_heavy_general(item, weight, mark);
+    }
+  }
+}
+
+template<typename T, typename S, typename A>
+void var_opt_sketch<T,S,A>::update_warmup_phase(const T& item, double weight, bool mark) {
+  assert(r_ == 0);
+  assert(m_ == 0);
+  assert(h_ <= k_);
+
+  if (h_ >= curr_items_alloc_) {
+    grow_data_arrays();
+  }
+
+  // store items as they come in until full
+  A().construct(&data_[h_], T(item));
+  weights_[h_] = weight;
+  if (marks_ != nullptr) {
+    marks_[h_] = mark;
+  }
+  ++h_;
+  num_marks_in_h_ += mark ? 1 : 0;
+
+  // check if need to heapify
+  if (h_ > k_) {
+    transition_from_warmup();
+  }
+}
+
+/* In the "light" case the new item has weight <= old_tau, so
+   would appear to the right of the R items in a hypothetical reverse-sorted
+   list. It is easy to prove that it is light enough to be part of this
+   round's downsampling */
+template<typename T, typename S, typename A>
+void var_opt_sketch<T,S,A>::update_light(const T& item, double weight, bool mark) {
+  assert(r_ >= 1);
```

Comment:
I believe we had an agreement to throw std::logic_error or something

Reply:
from (admittedly limited) conversations with people elsewhere, i suspect we're likely going to need to remove exceptions to make the library be considered more production-worthy

Reply:
Maybe, but asserts, if enabled, lead to exit with no way to recover, which is not production-worthy either. Perhaps, we need to bring this up on the dev list. I don't mind leaving this as it is for now.